### PR TITLE
add support of config doc sections

### LINKED ISSUE #4045: Support config doc sections
@machi1990 moving the discussion here.

I think we could just have:
```
    /**
     * Database related configuration.
     */
    @ConfigItem
    @ConfigDocSection
    public HibernateOrmConfigDatabase database;
```
because we will need a comment (it's mandatory) and we could have the first sentence of the comment be the title (we could automatically remove the dot) and the rest of the comment would be the intro of the section.

We could imagine supporting Asciidoc in this case too with `@asciidoclet`.

WDYT? Does it seem feasible?

### PR DESCRIPTION
**Pull request by machi1990**
Fixes #4045

/cc @gwenneg can you give this a spin and see if it satisfies your use case? 

/cc @gsmet

### CODE REVIEW

**Comment by gwenneg** [comment]
@machi1990 Sure, I'll give it a try ASAP (probably tomorrow). Thanks for submitting this so quickly! :)

**Comment by gwenneg** [comment]
> 
> 
> @machi1990 Sure, I'll give it a try ASAP (probably tomorrow). Thanks for submitting this so quickly! :)

I couldn't test it yesterday as I said, sorry about that, but it'll be done by the end the week-end for sure!

**Comment by machi1990** [comment]
> > @machi1990 Sure, I'll give it a try ASAP (probably tomorrow). Thanks for submitting this so quickly! :)
> 
> I couldn't test it yesterday as I said, sorry about that, but it'll be done by the end the week-end for sure!

hey @gwenneg worry out. I'll be looking forward to the results of your test.

**Comment by gwenneg** [comment]
I just cherry-picked 4e66adda036955b51fb1891075f34b4129b3d52e and tested it from the #3394 branch. 

The doc file produced thanks to `@ConfigDocSection` is great. There are only a few differences with the manual version:

- In #3394, there's an introduction sentence (`Each of the Caffeine caches backing up...`) in the Caffeine configuration section. With the generated version, I'll probably have to move it out of the section, but I don't think it's a problem because the sentence could be applied to any cache provider.
- The generated version exposes an internal attribute naming which is not very understandable for a user reading the doc: `quarkus.cache.caffeine."<namespace>".initial-capacity`. In the manual version, I replaced `<namespace>` with `cache-name`.

I also didn't find a way to generate the doc for the configuration part not annotated with `@ConfigDocSection` alone. The `quarkus-cache-caffeine.adoc` file contains the configuration related to `@ConfigDocSection` and the `quarkus-cache.adoc` file contains both this configuration and the general extension configuration.

**Comment by machi1990** [comment]
@gwenneg wow, thanks. I am always impressed on how detailed you can get. Thanks man.

> I just cherry-picked [4e66add](https://github.com/quarkusio/quarkus/commit/4e66adda036955b51fb1891075f34b4129b3d52e) and tested it from the #3394 branch.
> 
> The doc file produced thanks to `@ConfigDocSection` is great. There are only a few differences with the manual version:

Good, we are heading in the right directions. 

> 
> * In #3394, there's an introduction sentence (`Each of the Caffeine caches backing up...`) in the Caffeine configuration section. With the generated version, I'll probably have to move it out of the section, but I don't think it's a problem because the sentence could be applied to any cache provider.
> * The generated version exposes an internal attribute naming which is not very understandable for a user reading the doc: `quarkus.cache.caffeine."<namespace>".initial-capacity`. In the manual version, I replaced `<namespace>` with `cache-name`.

This is because the "key" is picked from the config item field name. Renaming this in your case will fix the issue. 

```
/**
  * Namespace configuration.
  */
@ConfigItem(name = ConfigItem.PARENT)
Map<String, CaffeineNamespaceConfig> cacheName;
``` 

> I also didn't find a way to generate the doc for the configuration part not annotated with `@ConfigDocSection` alone. The `quarkus-cache-caffeine.adoc` file contains the configuration related to `@ConfigDocSection` and the `quarkus-cache.adoc` file contains both this configuration and the general extension configuration.

This is a good point. Maybe we should generate the rest of elements not annotated `@ConfigDocSection` in a separate doc? @gwenneg @gsmet WDYT?

**Comment by gwenneg** [comment]
> @gwenneg wow, thanks. I am always impressed on how detailed you can get. Thanks man.

And I'm impressed by all the great PRs you submitted these past months ;)

> This is because the "key" is picked from the config item field name. Renaming this in your case will fix the issue.
> 
> ```
> /**
>   * Namespace configuration.
>   */
> @ConfigItem(name = ConfigItem.PARENT)
> Map<String, CaffeineNamespaceConfig> cacheName;
> ```

I did try renaming the attribute while testing and it worked, indeed. But this means we have to use a doc-friendly Java attribute name while the real technical meaning can be different. In the example above, `cacheName` doesn't reflect the real meaning/role of the map. I'm not sure this is a real problem though. WDYT?

> This is a good point. Maybe we should generate the rest of elements not annotated `@ConfigDocSection` in a separate doc? @gwenneg @gsmet WDYT?

Yes, I think this will be mandatory if we want the entire extension configuration doc to be generated.

**Comment by machi1990** [comment]
> > This is a good point. Maybe we should generate the rest of elements not annotated `@ConfigDocSection` in a separate doc? @gwenneg @gsmet WDYT?
> 
> Yes, I think this will be mandatory if we want the entire extension configuration doc to be generated.

Hi @gwenneg so I forgot to give an update about this. General properties are not generated on a separate file `extension-name-general-config-items.adoc` if there is a of `@ConfigDocSection`. Looking forward to your thoughts on this.  

/cc @gsmet

**Comment by gwenneg** [comment]
I just added ccd889af1686d0b2f7cc8ee7bee1f99d47c45c8c to my previous test and now everything I need is being generated at the right place, that's great thanks! :)

The doc-friendly map naming is still not convincing me though. I'd love to have a Java attribute with a meaningful name (for maintainability's sake) **and** an understandable name from a documentation consumer point of view at the same time. Maybe this could be done by adding a new annotation to override the default naming policy in the generated doc? Or maybe I'm overthinking this... I'd like to hear what others think about it.

**Comment by machi1990** [comment]
@gwenneg 

> I just added [ccd889a](https://github.com/quarkusio/quarkus/commit/ccd889af1686d0b2f7cc8ee7bee1f99d47c45c8c) to my previous test and now everything I need is being generated at the right place, that's great thanks! :)
>

Great. You welcome.
 
> The doc-friendly map naming is still not convincing me though. I'd love to have a Java attribute with a meaningful name (for maintainability's sake) **and** an understandable name from a documentation consumer point of view at the same time. Maybe this could be done by adding a new annotation to override the default naming policy in the generated doc? 

I was still thinking about it and I think a `@MapDocKey("value")` is  good solution.
For the cache extension we could have something like 
```
/**
  * Namespace configuration.
  */
@ConfigItem(name = ConfigItem.PARENT)
@MapDocKey("cache-name")
Map<String, CaffeineNamespaceConfig> namespace;
```
Of course,  if it is not present, we could default to the current behaviour when the annotation is not present.  Thoughts?

> Or maybe I'm overthinking this... I'd like to hear what others think about it.

Having a user friendly key is indeed a good idea.

**Comment by gwenneg** [comment]
Yes, that would be great.

**Comment by machi1990** [comment]
> Yes, that would be great.

Hi @gwenneg I have added the annotation as discussed above.

**Comment by gsmet** [comment]
Having an annotation is a good idea but please make it `@ConfigDocMapKey` for consistency. And we would make all the annotations with this prefix.

**Comment by gsmet** [comment]
As for the rest, I’m not sure to understand. @gwenneg you want to split the sections yourselves?

I must admit I haven’t had the time to check this out yet.

File: docs/src/main/asciidoc/extension-authors-guide.adoc [hunk 1]
```diff
@@ -2141,6 +2141,27 @@ You cannot use open blocks (`--`) in your Asciidoctor documentation.
...
+    /**
+     * Namespace configuration.
+     */
+    @ConfigItem(name = ConfigItem.PARENT)
+    @MapDocKey("cache-name") <1>
+    Map<String, CaffeineNamespaceConfig> namespace;
+}
+----
+
+1. This will generate a configuration map key named `quarkus.some."cache-name"` instead of `quarkus.some."<namespace>"`.
```
Comment (gwenneg) [inline docs/src/main/asciidoc/extension-authors-guide.adoc]: It looks like `<` and `>` should be removed from `<namespace>` here.

**Comment by gwenneg** [comment]
@machi1990 I just tested the lastest version with 91b2dc433929997765942601b861f7285c82169b and now everything's perfect for me. I'll use this in #3394 as soon as it's merged into `master`. Thanks!

@gsmet It all started from https://github.com/quarkusio/quarkus/pull/3394#issuecomment-531380583 and @machi1990 did a really nice job allowing me to separate the general caching config doc from the provider-specific config doc. We had a few iterations in this PR to fine-tune the initial version.

**Comment by gsmet** [comment]
Yeah, what I'm wondering is how we should split the sections.

Typically, in the case of Hibernate ORM, I would like to have separate sections but I want them all included in one `.adoc` because I don't want to manually add them.

I haven't taken a look at the PR yet but I'm wondering how it works for this and also how do we split the sections: do we have a summary for all of them then the details for all of them or summary + detail for each of them? TBH, I can't decide what's best.

**Comment by gwenneg** [comment]
In the current version of this PR, you have to manually include the sections into the extension `.adoc` which I believe is nice because it gives us the possibility to add some text in the doc between the sections. In the cache extension case for example, it could be useful to add a few lines about each provider on top of its configuration. I'm not sure to see how grouping all the summaries and then all the details would improve the doc. Keeping a summary and its details together seems more intuitive to me.

**Comment by machi1990** [comment]
> Yeah, what I'm wondering is how we should split the sections.
> 
> Typically, in the case of Hibernate ORM, I would like to have separate sections but I want them all included in one `.adoc` because I don't want to manually add them.

The way it is implemented right now is we generate a section file for each `ConfigDocSection` thus we'll have to manually include the sections files. 

> I haven't taken a look at the PR yet but I'm wondering how it works for this and also how do we split the sections: do we have a summary for all of them then the details for all of them or summary + detail for each of them? TBH, I can't decide what's best.

We have a summary and detail for all of them, just like the main generated doc.

**Comment by gsmet** [comment]
> In the current version of this PR, you have to manually include the sections into the extension .adoc which I believe is nice because it gives us the possibility to add some text in the doc between the sections.

My idea for that was to:
* use the first line of the ConfigGroup as the title (and remove the final dot if there is one)
* use the rest of the doc for a potential introduction - note that we would support Asciidoc too

I think it's the best of both world: we extract as much information as possible from the Javadoc. You don't forget adding a group to your documentation.

I agree it's a bit less flexible but I think it's a better approach, at least in the general case. I could be convinced that generating the full doc then the general + section documents if we really need this but we should have a very good reason to do so.

> Keeping a summary and its details together seems more intuitive to me.

But then you cannot have a quick overview of all the config properties, which is why we added the table initially :).

When I said, I couldn't decide, there are both pros and cons.

**Comment by machi1990** [comment]
> In the current version of this PR, you have to manually include the sections into the extension `.adoc` which I believe is nice because it gives us the possibility to add some text in the doc between the sections. In the cache extension case for example, it could be useful to add a few lines about each provider on top of its configuration. 

Now thinking about what Guillaume just said up there, I think if the javadoc is rich enough on itself, it will be great if we could generate the sections as part of the main doc that we way we would have to include the sections manually and preserve the ordering too.

**Comment by gsmet** [comment]
Maybe we should write something up in an issue next time we have a long discussion, @machi1990 so that we have a full summary of all our ideas :).

**Comment by gwenneg** [comment]
> My idea for that was to:
> 
> * use the first line of the ConfigGroup as the title (and remove the final dot if there is one)
> * use the rest of the doc for a potential introduction - note that we would support Asciidoc too
> 
> I think it's the best of both world: we extract as much information as possible from the Javadoc. You don't forget adding a group to your documentation.
> 
> I agree it's a bit less flexible but I think it's a better approach, at least in the general case. I could be convinced that generating the full doc then the general + section documents if we really need this but we should have a very good reason to do so.

I didn't think about it that way, but I like the idea and it would be enough for most (if not all) cases.

+1 for this!

**Comment by machi1990** [comment]
> Maybe we should write something up in an issue next time we have a long discussion, @machi1990 so that we have a full summary of all our ideas :).

:-) Yes, good to gather all the use cases. Actually my mind was more focused on the first TODO item of this issue https://github.com/quarkusio/quarkus/issues/3803 which somehow corresponds to @gwenneg use case.

**Comment by machi1990** [comment]
> > In the current version of this PR, you have to manually include the sections into the extension .adoc which I believe is nice because it gives us the possibility to add some text in the doc between the sections.
> 
> My idea for that was to:
> 
> * use the first line of the javadoc of the ConfigGroup as the title (and remove the final dot if there is one)
> * use the rest of the javadoc for a potential introduction - note that we would support Asciidoc too
> 
> I think it's the best of both world: we extract as much information as possible from the Javadoc. You don't forget adding a group to your documentation.

With the initial implementation, we could extract all this section information (title and introduction). What was missing is that this information was not included automatically in the main doc - i.e a separate section documentation file was created (part of issue #3803 which I have decided to leave aside) containing the information present on the java doc as introduction(at the top of the file, just above the summary table). 

> But then you cannot have a quick overview of all the config properties, which is why we added the table initially :).
> 

Yes, I totally that this is downside to that initial implementation. 

> > In the current version of this PR, you have to manually include the sections into the extension `.adoc` which I believe is nice because it gives us the possibility to add some text in the doc between the sections. In the cache extension case for example, it could be useful to add a few lines about each provider on top of its configuration.
> 
> Now thinking about what Guillaume just said up there, I think if the javadoc is rich enough on itself, it will be great if we could generate the sections as part of the main doc that we way we would not have to include the sections manually and preserve the ordering too.


With my recent push, I have changed things a bit and include sections in the main doc (in the details part) - e.g https://github.com/machi1990/quarkus-generated-extension-config-docs/blob/master/src/main/asciidoc/generated/quarkus-hibernate-orm.adoc (sections introduction are missing some meat, we'll get there :-)). 
@gwenneg do you mind trying this version out and see how it looks with the extension? 

With sections in play, I have change the ordering a bit so that sections (and its config items) appear at the end - see https://github.com/quarkusio/quarkus/pull/4050/files#diff-d0eef419555f654d57abf909ac9a0d62R92. 

@gwenneg @gsmet WDYT?

**Comment by gwenneg** [comment]
I should have some time to take a look at this and test the latest version tomorrow.

**Comment by FroMage** [comment]
I just tried your branch here, but I'm still getting the old-style docs before we did all the fixes, and no sections.

I'm not sure why.

**Comment by gsmet** [comment]
@machi1990 I gave it some thoughts in the train yesterday and here is a proposal. It's open to discussions as I'm not entirely sure it's the right solution so let's discuss it.

I think, I would do the following:

- no sections: same behavior as now
- sections, we get something like that:
```
= Reference documentation

(quick links to the sections)
 * General
 * Section 1

== General
=== Summary
The table
=== Details
The details

== Section 1
Intro
=== Summary
The table
=== Details
The details

== Section 2
Intro
=== Summary
The table
=== Details
The details
```

**Comment by gsmet** [comment]
@FroMage interested in your feedback on ^

The big question being: do we split per section or do we do all the summaries then all the details?

**Comment by FroMage** [comment]
Ah, the files in `docs/src/main/asciidoc/generated/` are old. I thought the build would update them.

**Comment by FroMage** [comment]
I guess I would have listed the sections in a single table, with section headers being a special row.

**Comment by gsmet** [comment]
@FroMage the full build is supposed to (if noone changed that). And if you build a specific extension, it just refreshes this extension.

**Comment by gsmet** [comment]
BTW, I started to integrate more guides the other day. I stumbled upon a Javadoc to Asciidoc issue that we need to fix but it's in the pipeline for 1.0.

**Comment by FroMage** [comment]
> the full build is supposed to (if noone changed that)

Looks like it did change, because my `generated` folder is never re-created.

**Comment by FroMage** [comment]
Oh, they now end up at the toplevel `target/asciidoc/generated/config/` huh

**Comment by gsmet** [comment]
@FroMage yes, there are 2 reasons:
- we were mixing generating resources and source code. David wasn't happy about it (and for a good reason).
- it wasn't possible to generate them in `docs/` because they are created before the `clean` for the `docs/` repo is made. So that's why they end up at the root, it's place where `clean` is made first so we can push things to `target`

**Comment by FroMage** [comment]
OK, but then there appears to be no build step that takes them there to put them to where they are rendered, that's an issue, no?

**Comment by FroMage** [comment]
Screenshot from 2019-10-04 11-23-00

That's what I got. I think it's alright.

**Comment by gsmet** [comment]
**Proposal 1** (the one from above)

```
= Reference documentation

(quick links to the sections)
 * General
 * Section 1

== General
=== Summary
The table
=== Details
The details

== Section 1
Intro
=== Summary
The table
=== Details
The details

== Section 2
Intro
=== Summary
The table
=== Details
The details
```

**Proposal 2**

```
= Reference documentation

== Summary
=== General
Intro
Table
=== Section 1
Intro
Table
=== Section 2
Intro
Table

== Details
=== General
Intro
Details
=== Section 1
Intro
Details
=== Section 2
Intro
Details
```

**Comment by gsmet** [comment]
@FroMage not sure I understand your issue. The files are copied to the website as part of the release process.

**Comment by FroMage** [comment]
Well when I do `mvn clean install` in the `docs` folder I get the rendered output. That's what I get to check the docs. Where the files currently are, they can't be looked at.

**Comment by gsmet** [comment]
@FroMage looks OK to me as long as we have the special rows. Having a link to the detailed section would be nice. We could use a Font Awesome icon for the link such as the plus in a circle or something.

BTW, please don't merge this PR as is as the split in section in the ORM doc needs some love and I need to review the code, too.

**Comment by gsmet** [comment]
@FroMage yes, they can. The includes are pointing to the root target directory (see `attributes.adoc`).

**Comment by FroMage** [comment]
> @FroMage yes, they can. The includes are pointing to the root target directory (see attributes.adoc).

Oh, OK, cool, then.

**Comment by FroMage** [comment]
@machi1990 can you add me as collaborator to your fork so I can push my commit?

I've fixed the section title, but did not add the link (not sure how to make an asciidoc link, I'm sure @machi1990 will know)

**Comment by FroMage** [comment]
Note that I'm still very unconvinced by the details section below because its formatting makes it incredibly hard to read. I don't know how to fix it though, and if not for searching I'd have made the main table expandable and gotten rid of it as it makes the page super long and will not help for the "all config" page.
Perhaps if we have some highlighting when we click on a key it would help. I'll look into that.

**Comment by machi1990** [comment]
@gsmet @FroMage thanks for your reviews. I have followed the discussion from a far since I am finishing something at work. I'll check this space a bit later with my feedbacks, but I like what Stéphane fixed. 

> @machi1990 can you add me as collaborator to your fork so I can push my commit?
> 
> I've fixed the section title, but did not add the link (not sure how to make an asciidoc link, I'm sure @machi1990 will know)

 I saw the image you shared about the fix, I like it. I usually check the `allow edits from maintainers` checkbox, I think with that you can force push the changes and it should be okay.

**Comment by machi1990** [comment]
> @FroMage looks OK to me as long as we have the special rows. Having a link to the detailed section would be nice. We could use a Font Awesome icon for the link such as the plus in a circle or something.

I can take a look at the link after @FroMage has pushed his recent changes.

> BTW, please don't merge this PR as is as the split in section in the ORM doc needs some love and I need to review the code, too.

Yes, the ORM doc needs some love. I'll put something if I have some inspirations, otherwise Guillaume you can go ahead and push something :-).

**Comment by FroMage** [comment]
> I usually check the allow edits from maintainers checkbox, I think with that you can force push the changes and it should be okay

Well it doesn't let me push to your branch, for some reason.

**Comment by FroMage** [comment]
@gsmet what's the plan for a all-config single-page? Given that we generate the docs per extension it doesn't look like we can generate docs for all the config in one go.

Best we can do is manually aggregate every page but they won't be sorted alphabetically, and they'll mix table/expanded which is very inelegant.

**Comment by FroMage** [comment]
> Well it doesn't let me push to your branch, for some reason.

sorry, it was my fault. I managed to push.

**Comment by gwenneg** [comment]
@machi1990 Sorry, I thought I'd have some time to test this again last week, but I ended up being absent from Quarkus for a week. I am now back. I'll catch up on this PR and test it again ASAP.

**Comment by machi1990** [comment]
This is superseded by #4381. Closing.

**Pull request closed without merging**